File: app/src/main/java/com/mx/gillustrated/component/MainActivityHeader.java
package com.mx.gillustrated.component;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.mx.gillustrated.R;
import com.mx.gillustrated.activity.MainActivity;
import com.mx.gillustrated.vo.CardInfo;

import java.util.HashMap;
import java.util.Map;

import butterknife.BindColor;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by maoxin on 2018/7/31.
 */

public class MainActivityHeader {

    private Map<String, TextView> mTextViewMap;
    private ListHeaderView mListHeaderView;
    private MainActivity mContext;
    private HeaderHandle mHeaderHandle;
    private ResourceController mResourceController;

    public MainActivityHeader(MainActivity context, HeaderHandle headerHandle, int gameId){
        mContext = context;
        mHeaderHandle = headerHandle;
        initialize();
        setResourceController(gameId);
    }

    public void setResourceController(int gameId){
        mResourceController = new ResourceController(mContext, gameId);
        mListHeaderView.tvHP.setText(mResourceController.getNumber1());
        mListHeaderView.tvAttack.setText(mResourceController.getNumber2());
        mListHeaderView.tvDefense.setText(mResourceController.getNumber3());
        mListHeaderView.tvExtra1.setText(mResourceController.getNumber4());
        mListHeaderView.tvExtra2.setText(mResourceController.getNumber5());
        if( "E1".equals(mResourceController.getNumber4())){
            mListHeaderView.ivExtraGap1.setVisibility(View.GONE);
            mListHeaderView.tvExtra1.setVisibility(View.GONE);
        }else{
            mListHeaderView.ivExtraGap1.setVisibility(View.VISIBLE);
            mListHeaderView.tvExtra1.setVisibility(View.VISIBLE);
        }
        if( "E2".equals(mResourceController.getNumber5())){
            mListHeaderView.ivExtraGap2.setVisibility(View.GONE);
            mListHeaderView.tvExtra2.setVisibility(View.GONE);
        }else{
            mListHeaderView.ivExtraGap2.setVisibility(View.VISIBLE);
            mListHeaderView.tvExtra2.setVisibility(View.VISIBLE);
        }
        if(!mContext.mSP.getBoolean(mContext.SHARE_SHOW_COST_COLUMN + gameId, false)){
            mListHeaderView.ivCostGap.setVisibility(View.GONE);
            mListHeaderView.tvCost.setVisibility(View.GONE);
        }else{
            mListHeaderView.ivCostGap.setVisibility(View.VISIBLE);
            mListHeaderView.tvCost.setVisibility(View.VISIBLE);
        }
    }

    private void initialize(){
        mListHeaderView = new ListHeaderView();
        setHeaderClickHandler(mListHeaderView.tvHP, CardInfo.COLUMN_MAXHP);
        setHeaderClickHandler(mListHeaderView.tvAttack, CardInfo.COLUMN_MAXATTACK);
        setHeaderClickHandler(mListHeaderView.tvDefense, CardInfo.COLUMN_MAXDEFENSE);
        setHeaderClickHandler(mListHeaderView.tvExtra1, CardInfo.COLUMN_EXTRA_VALUE1);
        setHeaderClickHandler(mListHeaderView.tvExtra2, CardInfo.COLUMN_EXTRA_VALUE2);
        setHeaderClickHandler(mListHeaderView.tvName, CardInfo.COLUMN_NAME);
        setHeaderClickHandler(mListHeaderView.tvAttr, CardInfo.COLUMN_ATTR);
        setHeaderClickHandler(mListHeaderView.tvCost, CardInfo.COLUMN_COST);
        setHeaderClickHandler(mListHeaderView.tvImg, CardInfo.COLUMN_NID);

        mTextViewMap = new HashMap<String, TextView>(){{
            put(CardInfo.COLUMN_NID, mListHeaderView.tvImg);
            put(CardInfo.COLUMN_MAXHP, mListHeaderView.tvHP);
            put(CardInfo.COLUMN_MAXATTACK, mListHeaderView.tvAttack);
            put(CardInfo.COLUMN_MAXDEFENSE, mListHeaderView.tvDefense);
            put(CardInfo.COLUMN_EXTRA_VALUE1, mListHeaderView.tvExtra1);
            put(CardInfo.COLUMN_EXTRA_VALUE2, mListHeaderView.tvExtra2);
            put(CardInfo.COLUMN_NAME, mListHeaderView.tvName);
            put(CardInfo.COLUMN_ATTR, mListHeaderView.tvAttr);
            put(CardInfo.COLUMN_COST, mListHeaderView.tvCost);
        }};

    }

    private void setHeaderClickHandler(TextView tv, final String column){
        tv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setHeaderColor(column);
                mHeaderHandle.onHeaderClick(column);
            }
        });
    }

    public void setHeaderColor(String orderby){
        for(Map.Entry<String, TextView> iterator : mTextViewMap.entrySet()){
            if(iterator.getValue() != null)
                iterator.getValue().setBackgroundColor(mListHeaderView.colorWhite2);
        }
        if(mTextViewMap.get(orderby) != null)
            mTextViewMap.get(orderby).setBackgroundColor(mListHeaderView.colorWhite);
    }

    class ListHeaderView{
        @BindView(R.id.tvHeaderImg)
        TextView tvImg;
        @BindView(R.id.tvHeaderName) TextView tvName;
        @BindView(R.id.tvHeaderAttr) TextView tvAttr;
        @BindView(R.id.tvHeaderCost) TextView tvCost;
        @BindView(R.id.tvHeaderHP) TextView tvHP;
        @BindView(R.id.tvHeaderAttack) TextView tvAttack;
        @BindView(R.id.tvHeaderDefense) TextView tvDefense;
        @BindView(R.id.tvHeaderExtra1) TextView tvExtra1;
        @BindView(R.id.tvHeaderExtra2) TextView tvExtra2;
        @BindView(R.id.ivExtra1Gap) ImageView ivExtraGap1;
        @BindView(R.id.ivExtra2Gap) ImageView ivExtraGap2;
        @BindView(R.id.ivCostGap) ImageView ivCostGap;
        @BindColor(R.color.color_white2) int colorWhite2;
        @BindColor(R.color.color_white) int colorWhite;

        public ListHeaderView()
        {
            View view = mContext.findViewById(R.id.ll_header);
            ButterKnife.bind(this, view);
        }
    }

    public interface HeaderHandle{
        void onHeaderClick(String c);
    }
}

File: app/src/main/java/com/mx/gillustrated/component/MainActivityHeader.kt
package com.mx.gillustrated.component

import android.annotation.SuppressLint
import android.graphics.Color
import android.view.View
import android.widget.TextView
import com.mx.gillustrated.R
import com.mx.gillustrated.activity.MainActivity
import com.mx.gillustrated.vo.CardInfo
import java.util.HashMap
import com.mx.gillustrated.activity.BaseActivity.Companion.SHARE_SHOW_COST_COLUMN
import com.mx.gillustrated.databinding.AdapterMainlistHeaderBinding

/**
 * Created by maoxin on 2018/7/31.
 */

class MainActivityHeader(private val mContext: MainActivity, private val mHeaderHandle: HeaderHandle, gameId: Int) {

    private lateinit var mTextViewMap: Map<String, TextView>
    private lateinit var mResourceController: ResourceController
    private var mHeaderPressColor:Int = Color.TRANSPARENT
    private var mHeaderDefaultColor:Int = Color.WHITE
    private lateinit var binding:AdapterMainlistHeaderBinding

    init {
        initialize()
        setResourceController(gameId)
    }

    fun setResourceController(gameId: Int) {
        mResourceController = ResourceController(mContext, gameId)
        binding.tvHeaderHP.text = mResourceController.number1
        binding.tvHeaderAttack.text = mResourceController.number2
        binding.tvHeaderDefense.text = mResourceController.number3
        binding.tvHeaderExtra1.text = mResourceController.number4
        binding.tvHeaderExtra2.text = mResourceController.number5
        if ("E1" == mResourceController.number4) {
            binding.ivExtra1Gap.visibility = View.GONE
            binding.tvHeaderExtra1.visibility = View.GONE
        } else {
            binding.ivExtra1Gap.visibility = View.VISIBLE
            binding.tvHeaderExtra1.visibility = View.VISIBLE
        }
        if ("E2" == mResourceController.number5) {
            binding.ivExtra2Gap.visibility = View.GONE
            binding.tvHeaderExtra2.visibility = View.GONE
        } else {
            binding.ivExtra2Gap.visibility = View.VISIBLE
            binding.tvHeaderExtra2.visibility = View.VISIBLE
        }
        if (!mContext.mSP.getBoolean(SHARE_SHOW_COST_COLUMN + gameId, false)) {
            binding.ivCostGap.visibility = View.GONE
            binding.tvHeaderCost.visibility = View.GONE
        } else {
            binding.ivCostGap.visibility = View.VISIBLE
            binding.tvHeaderCost.visibility = View.VISIBLE
        }
    }
    @SuppressLint("ResourceType")
    private fun initialize() {
        val attribute = intArrayOf(R.attr.colorGridPrimary, R.attr.colorBackPrimary)
        val array = mContext.theme.obtainStyledAttributes(attribute)
        mHeaderPressColor = array.getColor(0, Color.TRANSPARENT)
        mHeaderDefaultColor = array.getColor(1, Color.WHITE)
        array.recycle()

        binding =  mContext.binding.llHeaderInclude
        setHeaderClickHandler(binding.tvHeaderHP, CardInfo.COLUMN_MAXHP)
        setHeaderClickHandler(binding.tvHeaderAttack, CardInfo.COLUMN_MAXATTACK)
        setHeaderClickHandler(binding.tvHeaderDefense, CardInfo.COLUMN_MAXDEFENSE)
        setHeaderClickHandler(binding.tvHeaderExtra1, CardInfo.COLUMN_EXTRA_VALUE1)
        setHeaderClickHandler(binding.tvHeaderExtra2, CardInfo.COLUMN_EXTRA_VALUE2)
        setHeaderClickHandler(binding.tvHeaderName, CardInfo.COLUMN_NAME)
        setHeaderClickHandler(binding.tvHeaderAttr, CardInfo.COLUMN_ATTR)
        setHeaderClickHandler(binding.tvHeaderCost, CardInfo.COLUMN_COST)
        setHeaderClickHandler(binding.tvHeaderImg, CardInfo.COLUMN_NID)

        mTextViewMap = object : HashMap<String, TextView>() {
            init {
                put(CardInfo.COLUMN_NID, binding.tvHeaderImg)
                put(CardInfo.COLUMN_MAXHP, binding.tvHeaderHP)
                put(CardInfo.COLUMN_MAXATTACK, binding.tvHeaderAttack)
                put(CardInfo.COLUMN_MAXDEFENSE, binding.tvHeaderDefense)
                put(CardInfo.COLUMN_EXTRA_VALUE1, binding.tvHeaderExtra1)
                put(CardInfo.COLUMN_EXTRA_VALUE2, binding.tvHeaderExtra2)
                put(CardInfo.COLUMN_NAME, binding.tvHeaderName)
                put(CardInfo.COLUMN_ATTR, binding.tvHeaderAttr)
                put(CardInfo.COLUMN_COST, binding.tvHeaderCost)
            }
        }
    }

    private fun setHeaderClickHandler(tv: TextView, column: String) {
        tv.setOnClickListener {
            setHeaderColor(column)
            mHeaderHandle.onHeaderClick(column)
        }
    }

    fun setHeaderColor(orderby: String) {
        for ((_, value) in mTextViewMap) {
            value.setBackgroundColor(mHeaderPressColor)
        }
        if (mTextViewMap[orderby] != null)
            mTextViewMap[orderby]?.setBackgroundColor(mHeaderDefaultColor)
    }

    interface HeaderHandle {
        fun onHeaderClick(c: String)
    }
}
